From: Sunil Mushran <sunil.mushran@oracle.com>
Date: Wed, 22 Oct 2008 13:24:29 -0700
Subject: ocfs2: Set journal descriptor to NULL after journal shutdown
References: BZ 450579
Patch-mainline: 2.6.28

Patch sets journal descriptor to NULL after the journal is shutdown.
This ensures that jbd2_journal_release_jbd_inode(), which removes the
jbd2 inode from txn lists, can be called safely from ocfs2_clear_inode()
even after the journal has been shutdown.

Signed-off-by: Sunil Mushran <sunil.mushran@oracle.com>
Signed-off-by: Joel Becker <joel.becker@oracle.com>
Signed-off-by: Mark Fasheh <mfasheh@suse.com>
---
 fs/ocfs2/inode.c   |    6 ++++++
 fs/ocfs2/journal.c |    1 +
 2 files changed, 7 insertions(+), 0 deletions(-)

Index: linux-2.6.27-ocfs2/fs/ocfs2/inode.c
===================================================================
--- linux-2.6.27-ocfs2.orig/fs/ocfs2/inode.c
+++ linux-2.6.27-ocfs2/fs/ocfs2/inode.c
@@ -1106,6 +1106,12 @@ void ocfs2_clear_inode(struct inode *ino
 	oi->ip_last_trans = 0;
 	oi->ip_dir_start_lookup = 0;
 	oi->ip_blkno = 0ULL;
+
+	/*
+	 * ip_jinode is used to track txns against this inode. We ensure that
+	 * the journal is flushed before journal shutdown. Thus it is safe to
+	 * have inodes get cleaned up after journal shutdown.
+	 */
 	jbd2_journal_release_jbd_inode(OCFS2_SB(inode->i_sb)->journal->j_journal,
 				       &oi->ip_jinode);
 
Index: linux-2.6.27-ocfs2/fs/ocfs2/journal.c
===================================================================
--- linux-2.6.27-ocfs2.orig/fs/ocfs2/journal.c
+++ linux-2.6.27-ocfs2/fs/ocfs2/journal.c
@@ -690,6 +690,7 @@ void ocfs2_journal_shutdown(struct ocfs2
 
 	/* Shutdown the kernel journal system */
 	jbd2_journal_destroy(journal->j_journal);
+	journal->j_journal = NULL;
 
 	OCFS2_I(inode)->ip_open_count--;
 
